File: app/src/main/java/com/unidy2002/thuinfo/data/lib/Encrypt.java
package com.unidy2002.thuinfo.data.lib;

import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;
import android.util.Log;

import javax.crypto.*;
import javax.crypto.spec.GCMParameterSpec;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.*;

/*
/**
 * @author wutongke
 * @see "https://github.com/wutongke/KeyStoreDemo"
public class Encrypt {

    private static final String TRANSFORMATION = "AES/GCM/NoPadding";
    private static final String ANDROID_KEY_STORE = "AndroidKeyStore";

    private static Cipher cipherEncrypt;
    private static Cipher cipherDecrypt;
    private static KeyStore keyStore;

    public Encrypt(final String alias) throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException, InvalidAlgorithmParameterException, InvalidKeyException, UnrecoverableEntryException, KeyStoreException, java.security.cert.CertificateException, IOException {
        cipherEncrypt = Cipher.getInstance(TRANSFORMATION);
        cipherEncrypt.init(Cipher.ENCRYPT_MODE, getSecretKeyForEncrypt(alias));

        keyStore = KeyStore.getInstance(ANDROID_KEY_STORE);
        keyStore.load(null);

        final GCMParameterSpec spec = new GCMParameterSpec(128, cipherEncrypt.getIV());
        cipherDecrypt = Cipher.getInstance(TRANSFORMATION);
        cipherDecrypt.init(Cipher.DECRYPT_MODE, getSecretKeyForDecrypt(alias), spec);
    }

    public byte[] encrypt(final String text) throws BadPaddingException, IllegalBlockSizeException {
        return cipherEncrypt.doFinal(text.getBytes(StandardCharsets.UTF_8));
    }

    public String decrypt(final byte[] data) throws BadPaddingException, IllegalBlockSizeException {
        return new String(cipherDecrypt.doFinal(data), StandardCharsets.UTF_8);
    }

    private SecretKey getSecretKeyForEncrypt(final String alias) throws NoSuchProviderException, NoSuchAlgorithmException, InvalidAlgorithmParameterException {
        final KeyGenerator keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEY_STORE);
        keyGenerator.init(new KeyGenParameterSpec.Builder(alias,
                KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .build());
        return keyGenerator.generateKey();
    }

    private SecretKey getSecretKeyForDecrypt(final String alias) throws NoSuchAlgorithmException, UnrecoverableEntryException, KeyStoreException {
        return ((KeyStore.SecretKeyEntry) keyStore.getEntry(alias, null)).getSecretKey();
    }
}*/

File: app/src/main/java/com/unidy2002/thuinfo/data/util/Encrypt.kt
package com.unidy2002.thuinfo.data.util

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties.*
import java.nio.charset.StandardCharsets
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.spec.GCMParameterSpec

private const val KEY_ALGORITHM = KEY_ALGORITHM_AES
private const val BLOCK_MODE = BLOCK_MODE_GCM
private const val ENCRYPTION_PADDING = ENCRYPTION_PADDING_NONE
private const val TRANSFORMATION = "AES/GCM/NoPadding"
private const val ANDROID_KEY_STORE = "AndroidKeyStore"

/**
 * @see <a href="https://github.com/wutongke/KeyStoreDemo">KeyStoreDemo</a>
 * @see <a href="https://blog.csdn.net/pcaxb/article/details/47112245">ByteArray and String</a>
 */

fun encrypt(alias: String, text: String) =
    Cipher.getInstance(TRANSFORMATION).run {
        init(Cipher.ENCRYPT_MODE,
            KeyGenerator.getInstance(
                KEY_ALGORITHM,
                ANDROID_KEY_STORE
            ).run {
                init(
                    KeyGenParameterSpec.Builder(alias, PURPOSE_ENCRYPT or PURPOSE_DECRYPT)
                        .setBlockModes(BLOCK_MODE)
                        .setEncryptionPaddings(ENCRYPTION_PADDING)
                        .build()
                )
                generateKey()
            })
        b2s(iv) to b2s(doFinal(text.toByteArray(StandardCharsets.UTF_8)))
    }

fun decrypt(alias: String, key: Pair<String, String>) =
    Cipher.getInstance(TRANSFORMATION).run {
        init(Cipher.DECRYPT_MODE,
            (KeyStore.getInstance(ANDROID_KEY_STORE).run {
                load(null)
                getEntry(alias, null)
            } as KeyStore.SecretKeyEntry).secretKey,
            GCMParameterSpec(128, s2b(key.first))
        )
        String(doFinal(s2b(key.second)), StandardCharsets.UTF_8)
    }

private fun b2s(byteArray: ByteArray) =
    byteArray.joinToString("") { Integer.toHexString((0xFF and it.toInt())).padStart(2, '0') }

private fun s2b(string: String) =
    string.windowed(2, 2) { it.toString().toInt(16).toByte() }.toByteArray()